plugins/usbdmx/NodleU1.cpp

```diff
@@ -0,0 +1,538 @@
+    m_packet = new uint8_t[DATABLOCK_SIZE];
+    if (handle) {
+      SetInterfaceMode(m_adaptor, handle, m_mode);
+    }
+    return handle;
+  }
+
+  bool PerformTransfer(const DmxBuffer &buffer);
+
+  void PostTransferHook();
+
+ private:
+  unsigned int m_mode;
+  DmxBuffer m_tx_buffer;
+  // This tracks where we are in m_tx_buffer. A value of 0 means we're at the
+  // start of a DMX frame.
+  unsigned int m_buffer_offset;
+  uint8_t *m_packet;
+
+  bool ContinueTransfer();
+
+  bool SendInitialChunk(const DmxBuffer &buffer);
+
+  bool SendChunk() {
+    FillInterruptTransfer(WRITE_ENDPOINT, m_packet,
+                          DATABLOCK_SIZE, URB_TIMEOUT_MS);
+    return (SubmitTransfer() == 0);
+  }
+
+  DISALLOW_COPY_AND_ASSIGN(NodleU1AsyncUsbSender);
+};
+
+bool NodleU1AsyncUsbSender::PerformTransfer(const DmxBuffer &buffer) {
+  if (m_buffer_offset == 0) {
+    return SendInitialChunk(buffer);
+  }
+  // Otherwise we're part way through a transfer, do nothing.
+  return true;
+}
+
+void NodleU1AsyncUsbSender::PostTransferHook() {
+  if (m_buffer_offset < m_tx_buffer.Size()) {
+    ContinueTransfer();
+  } else if (m_buffer_offset >= m_tx_buffer.Size()) {
+    // That was the last chunk.
+    m_buffer_offset = 0;
+
+    if (TransferPending()) {
+      // If we have a pending transfer, the next chunk is going to be sent
+      // once we return.
+      m_tx_buffer.Reset();
+    }
+  }
+}
+
+bool NodleU1AsyncUsbSender::ContinueTransfer() {
+  unsigned int length = 32;
+
+  m_packet[0] = m_buffer_offset / 32;
+
+  m_tx_buffer.GetRange(m_buffer_offset, m_packet + 1, &length);
+  memset(m_packet + 1 + length, 0, 32 - length);
+  m_buffer_offset += length;
+  return (SendChunk() == 0);
+}
+
+bool NodleU1AsyncUsbSender::SendInitialChunk(const DmxBuffer &buffer) {
+  unsigned int length = 32;
+
+  m_tx_buffer.SetRange(0, buffer.GetRaw(), buffer.Size());
+
+  m_packet[0] = 0;
+  m_tx_buffer.GetRange(0, m_packet + 1, &length);
+  memset(m_packet + 1 + length, 0, 32 - length);
+
+  unsigned int slots_sent = length;
+  if (slots_sent < m_tx_buffer.Size()) {
+    // There are more frames to send.
+    m_buffer_offset = slots_sent;
+  }
+  return (SendChunk() == 0);
+}
+
+// AsynchronousNodleU1
+// -----------------------------------------------------------------------------
+
+AsynchronousNodleU1::AsynchronousNodleU1(
+    ola::usb::LibUsbAdaptor *adaptor,
+    libusb_device *usb_device,
+    PluginAdaptor *plugin_adaptor,
+    const string &serial,
+    unsigned int mode)
+    : NodleU1(adaptor, usb_device, plugin_adaptor, serial, mode) {
+  if (mode & 2) {  // output port active
+    m_sender.reset(new NodleU1AsyncUsbSender(m_adaptor, usb_device, mode));
+  }
+
+  if (mode & 4) {  // input port active
+    m_receiver.reset(new NodleU1AsyncUsbReceiver(m_adaptor, usb_device,
+                                                 plugin_adaptor, mode));
+  }
+}
+
+bool AsynchronousNodleU1::Init() {
+  bool err = false;
```

Comment:
It would be more readable if you wrote this as

bool ok = true;
ok &= m_sender->Init();
...
return ok;

That avoids all the negations.